File: shoring/compat/src/main/java/com/panopset/compat/FileProcessor.java
package com.panopset.compat;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

public class FileProcessor {
	private final File targetFile;
	private final boolean isRecursive;

	public FileProcessor(File file) {
		this(file, true);
	}

	public FileProcessor(File file, boolean isRecursive) {
		targetFile = file;
		this.isRecursive = isRecursive;
	}
	
	public FileProcessor(File file, LineFilter lineFilter) {
		this(file, lineFilter, true);
	}
	
	public FileProcessor(File file, LineFilter lineFilter, boolean isRecursive) {
		this(file, isRecursive);
		lineFilters.add(lineFilter);
	}

	private final List<LineFilter> lineFilters = new ArrayList<>();

	public FileProcessor withLineFilter(LineFilter lf) {
		lineFilters.add(lf);
		return this;
	}

	public boolean exec() {
		return new Report(targetFile, lineFilters).exec();
	}

	public void addFilters(List<LineFilter> filters) {
		if (filters == null || filters.isEmpty()) {
			return;
		}
		for (LineFilter lf : filters) {
			lineFilters.add(lf);
		}
	}
}

File: code/shoring/compat/src/main/kotlin/com/panopset/compat/FileProcessor.kt
package com.panopset.compat

import java.io.File
import java.util.*

class FileProcessor(private val targetFile: File, private val isRecursive: Boolean = true) {

    val transformer = Transformer()
    var byFileFilters: MutableList<ByFileFilter> = ArrayList()

    fun exec() {
        if (isRecursive) {
            recursiveExec( targetFile)
        } else {
            if (targetFile.isDirectory) {
                for (f in Objects.requireNonNull(targetFile.listFiles())) {
                    if (f.isFile) {
                        processFile( f)
                    }
                }
            } else {
                ReportFileTransformer(targetFile, transformer).exec()
            }
        }
    }

    private fun recursiveExec(rFile: File) {
        if (rFile.isDirectory) {
            for (f in Objects.requireNonNull(rFile.listFiles())) {
                recursiveExec( f)
            }
        } else {
            processFile( rFile)
        }
    }

    private fun processFile(rFile: File) {
        for (byFileFilter in byFileFilters) {
            if (!byFileFilter.fileFilter(rFile)) {
                return
            }
        }
        ReportFileTransformer(rFile, transformer).exec()
    }


    fun withFileFilter(byFileFilter: ByFileFilter): FileProcessor {
        byFileFilters.add(byFileFilter)
        return this
    }

    fun withLineFilter(byLineFilter: ByLineFilter): FileProcessor {
        transformer.byLineFilters.add(byLineFilter)
        return this
    }
}
